src/Tools/dotnet-monitor/DiagnosticsMonitorCommandHandler.cs

```diff
@@ -191,53 +202,61 @@ public IHostBuilder CreateHostBuilder(IConsole console, string[] urls, string[]
+                                if (address.Scheme.Equals(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase))
                                 {
                                     listenOptions.UseHttps();
                                 }
                             };
 
                             try
                             {
-                                if (url.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+                                if (address.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
                                 {
-                                    options.ListenLocalhost(url.Port, configureListenOptions);
+                                    options.ListenLocalhost(address.Port, configureListenOptions);
                                 }
-                                else if (IPAddress.TryParse(url.Host, out IPAddress ipAddress))
+                                else if (IPAddress.TryParse(address.Host, out IPAddress ipAddress))
                                 {
-                                    options.Listen(ipAddress, url.Port, configureListenOptions);
+                                    options.Listen(ipAddress, address.Port, configureListenOptions);
                                 }
                                 else
                                 {
-                                    options.ListenAnyIP(url.Port, configureListenOptions);
+                                    options.ListenAnyIP(address.Port, configureListenOptions);
                                 }
-                                boundListeningPort = true;
+                                bindingResults.AnyBoundPorts = true;
                             }
-                            catch (InvalidOperationException e)
+                            catch (InvalidOperationException ex)
                             {
-                                //This binding failure is typically due to missing default certificate
-                                console.Error.WriteLine($"Unable to bind to {url}. Dotnet-monitor functionality will be limited.");
-                                console.Error.WriteLine(e.Message);
+                                // This binding failure is typically due to missing default certificate.
+                                // Record the exception; it will be logged later through ILogger.
+                                bindingResults.Errors.Add(new AddressBindingResult(
+                                    $"Unable to bind to {url}. Dotnet-monitor functionality will be limited.",
+                                    ex));
                             }
                         }
-
-                        //If we end up not binding any ports, Kestrel defaults to port 5000. Make sure we don't attempt this.
-                        if (!boundListeningPort)
```

Comment:
Per the comment, the previous behavior I saw was that Kestrel used its default configuration (bind to 5000). This seems like the wrong thing to do.

Reply:
As described in Startup.Configure, Kestrel doesn't get a chance to startup because we throw an exception and stop application initialization.

Reply:
Ah I missed that, got it.